### PR TITLE
PCHR-1362 - Email notifications for documents

### PR DESCRIPTION
**Pull request by coldrunKacper**
Created 'senddocumentnotification' API action and BAO logic + Upgrader function installing daily scheduled job.

The new API action sends daily email notifications for documents including separate lists with:

- Overdue documents (!= approved)
- Due in next fortnight (14 days) (!= approved)
- Due in < 90 days (!= approved)

And show status in email.

Acceptance criteria:

- Emails are sent out to assignees only.
- Assignee contact receives lists of documents he/she is assigned to only
- Documents are grouped by due date according to the logic described in the task description

An exaple of email with documents notifications:
documents_notification

### CODE REVIEW

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Reminder.php [hunk 1]
```diff
@@ -486,6 +486,171 @@ private static function _getContactDailyReminderData($contactId, array $activity
...
+     * @return CRM_Core_DAO
+     */
+    protected static function queryDbForDocumentsNotificationData(DateTime $upToDate) {
+      $activityQuery = "SELECT a.id, a.activity_type_id, a.status_id, DATE(a.activity_date_time) AS activity_date,
+        GROUP_CONCAT(ac.record_type_id, ':', ac.contact_id, ':', contact.display_name, ':', e.email SEPARATOR  %1) AS activity_contact
+        FROM `civicrm_activity` a
+        LEFT JOIN civicrm_activity_contact ac ON ac.activity_id = a.id
+        LEFT JOIN civicrm_contact contact ON contact.id = ac.contact_id
+        LEFT JOIN civicrm_email e ON e.contact_id = ac.contact_id AND e.is_primary = 1
+        WHERE a.status_id <> 3 AND
```
Comment (omarabuhussein) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Reminder.php]: maybe defining **3** as constant is a better Idea and increase readability

File: uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Reminder.php [hunk 2]
```diff
@@ -486,6 +486,171 @@ private static function _getContactDailyReminderData($contactId, array $activity
...
+     * Return an array containing a set of Documents for each Assignee
+     * grouped by due date period ('overdue' - overdue, 'plus14' - due date
+     * within next 14 days, 'plus90' - due date within next 90 days).
+     * 
+     * @return array
+     */
+    protected static function getDocumentNotificationsData() {
+      $today = new DateTime();
+      $todayPlus14 = (new DateTime())->add(new DateInterval('P14D'));
+      $todayPlus90 = (new DateTime())->add(new DateInterval('P90D'));
```
Comment (omarabuhussein) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Reminder.php]: maybe you can make this more flexable by passing due dates as parameters to **getDocumentNotificationsData** method

Comment (coldrunKacper) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Reminder.php]: I think it's not necessary at this point - as we use 14 days and 90 days also in template + these two periods have their own 'string' description so defining it all as a parameter would require to specify more parameters than just two dates and then the method definition would become chaotic a bit.

Comment (omarabuhussein) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Reminder.php]: > as we use 14 days and 90 days also in template

you mean such as here https://github.com/compucorp/civihr-tasks-assignments/pull/96/files#diff-c7ad0da2fe38025e6536e37641e28784R21 ... I think you can pass dueDate1 and dueDate2 to smarty and use them there too ..

> these two periods have their own 'string' description
I am not getting this , can you show me an example ?

Comment (coldrunKacper) [inline uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Reminder.php]: "Documents due in next fortnight" - this is a description for 14 days variant in the template. Also, if I make the due dates as a parameter - it requires to assume that there are always two dates (for 14 and 90 days) - so it still will be not 100% flexible.

**Comment by omarabuhussein** [comment]
approved

**Pull request merged**